### PR TITLE
[WebNN] Improve data type check of slice op

### PR DESCRIPTION
**Pull request by shiyi9801**
A follow-up of [[WebNN] Support negative steps for slice](https://github.com/microsoft/onnxruntime/pull/22871#discussion_r1847929774). Slice op is emulated by reverse+slice when steps < 0 so `SliceOpBuilder::HasSupportedInputsImpl()` should also check the supported data types of reverse.

### CODE REVIEW

**Comment by shiyi9801** [comment]
@Honry PTAL, thanks!

File: onnxruntime/core/providers/webnn/builders/helper.cc [hunk 1]
```diff
@@ -177,14 +177,31 @@ bool IsDataTypeSupportedByOp(const std::string& onnx_op_type,
...
+                          << webnn_input_output_name << "]";
+    return false;
+  }
   if (!IsSupportedDataType(onnx_data_type, wnn_limits[webnn_op_type][webnn_input_output_name]["dataTypes"])) {
-    LOGS(logger, VERBOSE) << "[" << onnx_op_type
-                          << "] " << onnx_input_output_name
-                          << " type: [" << onnx_data_type
-                          << "] is not supported for now";
+    LOGS(logger, VERBOSE) << "[" << onnx_op_type << "] WebNN op [" << webnn_op_type << "] " << onnx_input_output_name
+                          << " data type: [" << onnx_data_type << "] is not supported by WebNN for now";
```
Comment (Honry) [inline onnxruntime/core/providers/webnn/builders/helper.cc:202]: How about this?
```suggestion
    LOGS(logger, VERBOSE) << "[" << onnx_op_type << "] " << onnx_input_output_name
                          << "'s data type: [" << onnx_data_type 
                          << "] is not supported by WebNN op ["
                          << webnn_op_type << "] for now";
```

**Review by Honry — COMMENTED** [review]
Thanks @shiyi9801, data type check is not rigorous in WebNN EP, especially for those ONNX ops that need to be supported by decomposed WebNN ops. This is a good starting.👍

File: onnxruntime/core/providers/webnn/builders/impl/slice_op_builder.cc [hunk 2]
```diff
@@ -161,6 +163,30 @@ bool SliceOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializers,
...
+  if (!GetType(input, input_type, logger))
+    return false;
+
+  // If there is step < 0, check data type support of reverse.
+  if (input_defs.size() > 4 && input_defs[4]->Exists()) {
+    std::vector<int64_t> steps;
+    if (!ReadIntArrayFrom1DTensor(*initializers.at(input_defs[4]->Name()), steps, logger))
+      return false;
+    if (std::any_of(steps.begin(), steps.end(), [](int64_t step) { return step < 0; })) {
+      if (!IsDataTypeSupportedByWebNNOp(op_type, "reverse", input_type, wnn_limits, "input", "data", logger)) {
```
[the remark this answers is not in the document]
Comment (shiyi9801) [reply onnxruntime/core/providers/webnn/builders/impl/slice_op_builder.cc:181]: `slice` always needs to be checked (except for a few cases like steps=1, start=0, end=input_dim) so we still have to call `IsDataTypeSupportedByWebNNOp()` twice.

[the remark this answers is not in the document]
Comment (Honry) [reply onnxruntime/core/providers/webnn/builders/impl/slice_op_builder.cc:181]: Oh, sorry, I missed this.

**Review by Honry — APPROVED** [review]
👍

**Comment by shiyi9801** [comment]
Hi @fdwr PTAL, thanks!

File: onnxruntime/core/providers/webnn/builders/impl/slice_op_builder.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (fdwr) [reply onnxruntime/core/providers/webnn/builders/impl/slice_op_builder.cc:181]: I wondered when `slice` and `reverse` would ever differ in WebNN data type support, as all data movement operators *should* support the same data types (and if they don't, that's very dubious). Though, looking at TFLiteRT, I now see unfortunate inconsistencies o_o.

**reverse**:
32-bit float
1-bit signless integer values
8-bit unsigned integer
16-bit signless integer <<<<
32-bit signless integer
64-bit signless integer
QUI8
QI8
QI16
quint8

**slice**:
32-bit float
1-bit signless integer
8-bit signless integer
32-bit signless integer
64-bit signless integer
8-bit unsigned integer
32-bit unsigned integer <<<<
TFLite string type <<<<
QI8
QUI8
QI16
quint8 type

I support the for general rigorousness.

**Review by fdwr — DISMISSED** [review]
👍

**Comment by fdwr** [comment]
The ~~bothersome~~ linter appears to complain about this line...

```
    LOGS(logger, VERBOSE) << "[" << onnx_op_type << "] " << onnx_input_output_name
                          << "'s data type: [" << onnx_data_type <------------------------
                          << "] is not supported by WebNN op ["
                          << webnn_op_type << "] for now";
```

...but I can't tell what's wrong o_o. https://github.com/microsoft/onnxruntime/actions/runs/12151097208/job/34052156380?pr=22988

**Comment by shiyi9801** [comment]
> The ~bothersome~ linter appears to complain about this line...
> 
> ```
>     LOGS(logger, VERBOSE) << "[" << onnx_op_type << "] " << onnx_input_output_name
>                           << "'s data type: [" << onnx_data_type <------------------------
>                           << "] is not supported by WebNN op ["
>                           << webnn_op_type << "] for now";
> ```
> 
> ...but I can't tell what's wrong o_o. https://github.com/microsoft/onnxruntime/actions/runs/12151097208/job/34052156380?pr=22988
> 
> All other tests passed.

There is a trailing whitespace

> << "'s data type: [" << onnx_data_type <------------------------

so the linter clang-format complained🥲 Fixed it, please help approve again, thanks! @fdwr

**Review by fdwr — APPROVED** [review]
👍

**Pull request merged**